File: forge/src/main/java/online/ebatel/forge/WebAdminMod.java
package online.ebatel.forge;

import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.RegisterCommandsEvent;
import net.minecraftforge.event.server.ServerStartingEvent;
import net.minecraftforge.event.server.ServerStoppingEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;
import online.ebatel.forge.commands.PlayerListCommand;
import online.ebatel.forge.commands.ServerStatsCommand;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@Mod(WebAdminMod.MOD_ID)
public class WebAdminMod {
    public static final String MOD_ID = "webadminplugin";
    public static final Logger LOGGER = LogManager.getLogger();

    private WebhookClient webhookClient;
    private ChatEventHandler chatEventHandler;

    public WebAdminMod() {
        ModConfig.register();
        MinecraftForge.EVENT_BUS.register(this);
        MinecraftForge.EVENT_BUS.register(TpsTracker.class);
    }

    @SubscribeEvent
    public void onServerStarting(ServerStartingEvent event) {
        LOGGER.info("WebAdminPlugin (Forge) enabled!");

        // Setup chat webhook
        if (ModConfig.WEBHOOK_ENABLED.get()) {
            String webhookUrl = ModConfig.getWebhookUrl();
            if (webhookUrl != null) {
                webhookClient = new WebhookClient(LOGGER);
                chatEventHandler = new ChatEventHandler(webhookClient);
                MinecraftForge.EVENT_BUS.register(chatEventHandler);
                LOGGER.info("Chat webhook enabled: {}", webhookUrl);
            } else {
                LOGGER.warn("Webhook URL is not configured. Chat forwarding disabled.");
            }
        }
    }

    @SubscribeEvent
    public void onServerStopping(ServerStoppingEvent event) {
        if (chatEventHandler != null) {
            MinecraftForge.EVENT_BUS.unregister(chatEventHandler);
        }
        if (webhookClient != null) {
            webhookClient.shutdown();
        }
        LOGGER.info("WebAdminPlugin (Forge) disabled!");
    }

    @SubscribeEvent
    public void onRegisterCommands(RegisterCommandsEvent event) {
        PlayerListCommand.register(event.getDispatcher());
        ServerStatsCommand.register(event.getDispatcher());
        LOGGER.info("WebAdminPlugin commands registered!");
    }
}

File: fabric/shared/src/main/kotlin/online/ebatel/fabric/WebAdminMod.kt
package online.ebatel.fabric

import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.command.v2.CommandRegistrationCallback
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents
import net.fabricmc.fabric.api.message.v1.ServerMessageEvents
import online.ebatel.fabric.commands.PlayerListCommand
import online.ebatel.fabric.commands.ServerStatsCommand
import org.slf4j.LoggerFactory

/**
 * Fabric entry point. Mirrors the behaviour of the Bukkit/Forge variants:
 *  - registers the /playerlist and /serverstat commands (consumed by the admin panel over RCON)
 *  - tracks TPS every server tick
 *  - optionally forwards in-game chat to the admin panel webhook
 *
 * Registered as a Kotlin entry point, so the server needs fabric-language-kotlin installed.
 */
object WebAdminMod : ModInitializer {
    const val MOD_ID = "webadminplugin"

    @JvmField
    val LOGGER = LoggerFactory.getLogger(MOD_ID)

    private var webhookClient: WebhookClient? = null

    override fun onInitialize() {
        FabricModConfig.load()

        // Keep TPS samples flowing regardless of webhook config (serverstat needs them).
        ServerTickEvents.END_SERVER_TICK.register(ServerTickEvents.EndTick { TpsTracker.onServerTick() })

        CommandRegistrationCallback.EVENT.register(
            CommandRegistrationCallback { dispatcher, _, _ ->
                PlayerListCommand.register(dispatcher)
                ServerStatsCommand.register(dispatcher)
            }
        )
        LOGGER.info("WebAdminPlugin commands registered!")

        val webhookUrl = FabricModConfig.getWebhookUrl()
        if (webhookUrl != null) {
            val client = WebhookClient(LOGGER)
            webhookClient = client
            ServerMessageEvents.CHAT_MESSAGE.register(
                ServerMessageEvents.ChatMessage { message, sender, _ ->
                    // Re-read each time: config may have been reloaded; cheap.
                    val url = FabricModConfig.getWebhookUrl() ?: return@ChatMessage
                    client.sendChatMessage(
                        url,
                        sender.name.string,
                        sender.uuid.toString(),
                        message.signedContent()
                    )
                }
            )
            LOGGER.info("Chat webhook enabled: {}", webhookUrl)
        } else {
            LOGGER.info("Chat webhook disabled (see config/{}.json).", MOD_ID)
        }

        ServerLifecycleEvents.SERVER_STOPPING.register(
            ServerLifecycleEvents.ServerStopping { webhookClient?.shutdown() }
        )

        LOGGER.info("WebAdminPlugin (Fabric) initialized!")
    }
}
